chronos/asyncsync.nim

```diff
@@ -638,3 +644,21 @@ proc waitEvents*[T](ab: AsyncEventQueue[T],
         break
 
   events
+
+proc newWaitGroup*(count: int): WaitGroup =
+  doAssert(count >= 0, "WaitGroup count must be non negative number")
+  let fut = newFuture[void]("WaitGroup")
+  if count == 0:
+    fut.complete()
+  WaitGroup(count: count, fut: fut)
+
+proc wait*(wg: WaitGroup): Future[void] =
+  wg.fut
```

Comment:
As an additional comment to this like you should return
```nim
  wg.fut.join()
```
instead of `fut` itself, because oherwise you transferring ownership of internal future which should be managed internally to outer world. So if somebody will cancel some operation which will be involved in `await waitGroup.wait()` this internal future will be cancelled as well, and so become actually finished.

Reply:
aha, got it! please have a look again